src/libraries/System.Runtime.InteropServices/src/System/Runtime/InteropServices/PosixSignalRegistration.Unix.cs

```diff
@@ -0,0 +1,173 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.ComponentModel;
+using System.Threading;
+
+namespace System.Runtime.InteropServices
+{
+    public sealed class PosixSignalRegistration : IDisposable
+    {
+        private readonly Action<PosixSignalContext> _handler;
+        private readonly PosixSignal _signal;
+        private bool _registered;
+        private readonly object _gate = new object();
+
+        private PosixSignalRegistration(PosixSignal signal, Action<PosixSignalContext> handler)
+        {
+            _signal = signal;
+            _handler = handler;
+        }
+
+        public static PosixSignalRegistration Create(PosixSignal signal, Action<PosixSignalContext> handler)
+        {
+            if (handler == null)
+            {
+                throw new ArgumentNullException(nameof(handler));
+            }
+            if (signal > PosixSignal.SIGHUP || signal < PosixSignal.SIGCHLD)
+            {
+                throw new IndexOutOfRangeException();
+            }
+            PosixSignalRegistration registration = new PosixSignalRegistration(signal, handler);
+            registration.RegisterFor(signal);
+            return registration;
+        }
+
+        private unsafe void RegisterFor(PosixSignal signal)
+        {
+            if (!s_initialized)
+            {
+                if (!Interop.Sys.InitializeTerminalAndSignalHandling())
```

Comment:
Please update:
https://github.com/dotnet/runtime/blob/01b7e73cd378145264a7cb7a09365b41ed42b240/src/libraries/Native/Unix/System.Native/pal_console.c#L476
as well.  Beyond being used by multiple libraries, the static public PosixSignal now depends on the lock for correctness.